### DISCUSSION TITLE
How does the Pivoted Cholesky Decomposition implementation differ from Harbrecht et.al. 2012 ?

### DISCUSSION THREAD

**Discussion by kunalghosh** [cornellius-gp/gpytorch#2067] [Q&A]
Hi GPyTorch devs, I am trying to implement the `mBCG` algorithm from the original GPyTorch paper for PyMC. I am currently working on implementing _Pivoted Cholesky decomposition_ and have implemented the one from the Original Harbrecht et.al 2012 paper

![Screenshot 2022-07-15 at 10 05 07 PM](https://user-images.githubusercontent.com/157300/179293916-8eb95797-9e19-46f2-b892-a7ad9b166614.png)

To me it seems like the [implementation in GPyTorch](https://github.com/cornellius-gp/gpytorch/blob/78796f7e3d36530fbb4a395cf9b303c680367c0d/gpytorch/functions/_pivoted_cholesky.py#L1) differs from the original in three places, 

1. The inner for-loop runs until `max_iter` or `rank` instead of `n` 
2. Initial error is `max(diag)` instead of `l1_norm(diag)`
3. Error computed in the loop is divided by the Initial error 

Are there any other differences ?

I wrote a simple test (with my Pivoted Cholesky implementation in NumPy) and the results differ from GPyTorch. [link to the notebook](https://colab.research.google.com/drive/1sLNdLi3sI0JKO9ooOsuS6aKFBDCdN3n8?usp=sharing)

Any help / pointers would be very helpful. (tagging @gpleiss since he implemented the algorithm on GPyTorch)

**Comment by gpleiss** [comment]
> The inner for-loop runs until max_iter or rank instead of n

You could fix this by setting `gpytorch.settings.max_preconditioner_size` to `n`. We have a maximum number of iterations because we are primarily using this function for preconditioning.

> Initial error is max(diag) instead of l1_norm(diag)

I don't think that's true? https://github.com/cornellius-gp/gpytorch/blob/78796f7e3d36530fbb4a395cf9b303c680367c0d/gpytorch/functions/_pivoted_cholesky.py#L32

> Error computed in the loop is divided by the Initial error

Most of our numerical methods in GPyTorch use relative errors, not absolute errors.

It sounds like the only difference between your implementation and ours is the stopping conditions. If you need strict stopping conditions based on absolute error, then our implementation probably isn't suitable for your needs.
